Code/Source/solver/mat_models.cpp

```diff
@@ -138,6 +250,57 @@ void get_fib_stress(const ComMod& com_mod, const CepMod& cep_mod, const fibStrsT
   }
 }
 
+
+/**
+ * @brief Perform the necessary tensor operations to calculate S_iso (isochoric
+ * 2nd PK stress) from its fictitious counterpart S_bar, and CC_iso (isochoric
+ * material elasticity tensor) from its fictitious counterpart CC_bar.
+ * 
+ * In particular, performs the following calculations:
+ * 
+ * S_iso = J^(-2/nsd) * P : S_bar
+ * where P is the 4th order projection tensor
+ * P = I - 1/3 * C^-1 ⊗ C
+ * where I is the 4th order identity tensor, C is the right Cauchy-Green tensor
+ * More efficiently, we can write
+ * S_iso = J^(-2/nsd) * S_bar - r1 * C^-1
+ * 
+ * CC_iso = P : CC_bar : P^T  
+ *        + 2/nsd * (C^-1 ⊗ S_iso + S_iso ⊗ C^-1) 
+ *        + 2 * r1 * sym(C^-1 ⊗ C^-1) - 2 * r1/nsd * (C^-1 ⊗ C^-1)
+ * 
+ * where r1 = J^(-2/nsd) * C : S_bar / nsd
+ * 
+ * Follows theory from "A General Approach to Derive Stress and Elasticity Tensors
+ * for Hyperelastic Isotropic and Anisotropic Materials" by Cheng and Zhang.
+ * 
+ */
+template<size_t nsd>
+std::pair<EigenMatrix<nsd>, EigenTensor<nsd>> bar_to_iso(
+  const EigenMatrix<nsd>& S_bar, const EigenTensor<nsd> &CC_bar, 
+  const double J2d, const EigenMatrix<nsd>& C, const EigenMatrix<nsd>& Ci) 
+  {
+
+  using namespace mat_fun;
+
+  // Useful scalar
+  double r1 = J2d * mat_ddot_eigen<nsd>(C, S_bar) / nsd;
+
+  // Compute isochoric 2nd Piola-Kirchhoff stress
+  auto S_iso = J2d*S_bar - r1*Ci;
+
+  // Compute isochoric material elasticity tensor
+  EigenTensor<nsd> PP = ten_ids_eigen<nsd>() - (1.0/nsd) * ten_dyad_prod_eigen<nsd>(Ci, C); // Important: using auto here causes tests to fail
+  auto CC_iso = ten_ddot_eigen<nsd>(CC_bar, PP);
+  CC_iso = ten_transpose_eigen<nsd>(CC_iso);
+  CC_iso = ten_ddot_eigen<nsd>(PP, CC_iso);
+  CC_iso += (-2.0/nsd) * (ten_dyad_prod_eigen<nsd>(Ci, S_iso)  + ten_dyad_prod_eigen<nsd>(S_iso, Ci));
+  CC_iso += 2.0 * r1 * ten_symm_prod_eigen<nsd>(Ci, Ci)  +  (- 2.0*r1/nsd) * ten_dyad_prod_eigen<nsd>(Ci, Ci);
+
```

Comment:
Would it be possible to replace some of the function calls by expressions ? 
```
auto CC_iso = CC_bar.contract(PP, contractionDims);
```
Would Eigen then be able to do a better job of optimizing them?

Reply:
Yes, in fact, currently `double_dot_product_2323()` is just a wrapper on the `contract()` method. I think it's just a question of readability. Here is the code contained inside `double_dot_product_2323()`

        // Define the contraction dimensions
        Eigen::array<Eigen::IndexPair<int>, 2> contractionDims = {
            Eigen::IndexPair<int>(2, 2), // Contract A's 2nd dimension with B's 2nd
            Eigen::IndexPair<int>(3, 3)  // Contract A's 3rd dimension with B's 3rd
        };

        // Return the double dot product
        return A.contract(B, contractionDims);

I feel that by wrapping this code and the definition of contractionDims, it is more readable. But open to suggestions!

Reply:
@aabrown100-git The `double_dot_product_2323()` and `double_dot_product_3424()` functions are only used in `_get_pk2cc()` so I would define `prod_dims2323` and `prod_dims3424` locally and 
```
CC = CC. contract(CC_bar, prod_dims2323);
```

Or  if these functions are going to be use more often then generalize `double_dot_product()` to use product dimensions as parameters
```
double_dot_product(CC, {2,3}, CC_bar, {2,3});
```

Reply:
Sounds good, I will go with the generalized `double_dot_product()`